src/material/datepicker/month-view.ts

```diff
@@ -329,7 +340,14 @@ export class MatMonthView<D> implements AfterContentInit, OnChanges, OnDestroy {
       this.activeDateChange.emit(this.activeDate);
     }
 
+    if (!event.isTrusted) {
+      // Manually triggered events in unit tests do not trigger change detection.
+      this._changeDetectorRef.detectChanges();
+    }
```

Comment:
This isn't really the right approach to this issue. The unit test should directly call `detectChanges`- library code (this code) shouldn't add additional behaviors explosively to make unit tests pass. Having the tests call `detectChanges` in tandem with emitting the fake event is the normal/standard way of handling this.

Reply:
Well, calling `detectChanges` in the unit test doesn't work in this situation because the microtask queue is flushing before change detection runs. the child component (`MatCalendarBody`) needs to received the updated data from `this.activeDate` before it can focus on the correct cell in this method https://github.com/angular/components/blob/194ec1b7aad7df4b3deed4820c983151d0bad978/src/material/datepicker/calendar-body.ts#L209.

Here's an explanation I wrote up earlier which seems to have gotten buried in the PR comments.

I just pushed a change to only call detectChanges if the event is not trusted. This is only needed for the tests because fake keyboard events do not trigger change detection. I'm generally not a fan of having code that's only intended to run in the tests, and I'm curious how we usually handle this situation.

Calling detectChanges in the test code doesn't work because by that time the calendar body will have already triggered the activeDateChange event with a stale value. This wasn't a problem until now, because before focusing on a stale date wouldn't trigger the activeDateChange event to update the active date. Calling detectChanges in the test code would case the calendar body to focus again but on a updated value.

Here's a rough timeline of what happens in the unit tests if we don't call detectChanges when handling the keydown event.
0. (active Date is Jan 12 2021)
1. trigger a fake keyboard event with the right arrow key
2. MatMonthView sets this.activeDate to Jan 13 2021
3. MatMonthView calls _focusActiveDate
4. MatCalendarBody calls .focus on Jan 12, 2021 date then triggers an activeDateChange Event with Jan 12, 2021
5. test code calls detectChanges
6. month-view receives sets the active date to Jan 12, 2021
7. (active date is Jan 12, but it should be Jan 13, 2021).

There is two approaches I can think of, but both required changing the library code (this code)
1. call detect changes in library code (this code) only if the event is not trusted
2. add a `timeOut` to `_focusActiveCell` so that the microtask queue does not get flushed until we manually flush it in the test code.

WDYT?

Reply:
If it is too difficult to handle testing these in unit tests, I would suggest just test these cases in our e2e tests with real keyboard events